compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -602,6 +603,44 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
         let ty = self.infcx.resolve_vars_if_possible(self.node_ty(var_hir_id));
 
+        let ty = match closure_clause {
+            hir::CaptureBy::Value => ty, // for move closure the capture kind should be by value
+            hir::CaptureBy::Ref => {
+                // Max capture kind according to Imm < UniqueImm < Mut < ByValue
+                let mut max_capture_kind =
+                    root_var_min_capture_list.first().unwrap().info.capture_kind;
+                for capture in root_var_min_capture_list.iter() {
+                    max_capture_kind = match (max_capture_kind, capture.info.capture_kind) {
```

Comment:
I'm assuming this is 1-1 copy of the code that's in another helper funciton? Can we abstract this chunk out to another helper and use in both function?

Reply:
Yes, I was gonna do that once I had confirmed this is what we wanted to do